facade/service.go

```diff
@@ -176,32 +176,90 @@ func (f *Facade) MigrateService(ctx datastore.Context, request dao.ServiceMigrat
 		}
 	}
 
-	svcs, err = readNewServiceDefinitions(outputFileName)
+	svcsMap, err := readNewServiceDefinitions(outputFileName)
 	if err != nil {
 		return err
 	}
 
-	for _, svc := range svcs {
-		if request.DryRun {
-			err = f.verifyServiceForUpdate(ctx, svc, nil)
-			if err == nil {
-				glog.V(2).Infof("Facade:MigrateService: dry-run of migration script complete for serviceID %+v", svc.ID)
+	// Validate the modified services.
+	for _, svc := range svcsMap["modified"] {
+		if err = f.verifyServiceForUpdate(ctx, svc, nil); err != nil {
+			return err
+		}
+	}
+
+	// Validate the cloned services.
+	existing, err := f.serviceStore.GetServices(ctx)
+	if err != nil {
+		return err
+	}
+	apps := map[string]bool{}
+	for _, svc := range existing {
+		for _, ep := range svc.Endpoints {
+			if ep.Purpose == "export" {
+				apps[ep.Application] = true
 			}
-		} else {
-			glog.V(2).Infof("Facade.MigrateService: migration script complete, updating serviceID %+v", svc.ID)
-			err := f.stopServiceForUpdate(ctx, *svc)
-			if err != nil {
-				return err
+		}
+	}
+	for _, svc := range svcsMap["cloned"] {
+		svc.ID, err = utils.NewUUID36()
+		if err != nil {
+			return err
+		}
+		now := time.Now()
+		svc.CreatedAt = now
+		svc.UpdatedAt = now
+
+		// verify the service with name and parent does not collide with another existing service
+		if s, err := f.serviceStore.FindChildService(ctx, svc.DeploymentID, svc.ParentServiceID, svc.Name); err != nil {
+			return err
+		} else if s != nil {
+			if s.ID != svc.ID {
+				return fmt.Errorf("Duplicate name detected for service %s found at %s", svc.Name, svc.ParentServiceID)
 			}
-			err = f.migrateService(ctx, svc)
 		}
 
-		if err != nil {
-			break
+		// Make sure no endpoint apps are duplicated.
+		for _, ep := range svc.Endpoints {
+			if ep.Purpose == "export" {
+				if _, ok := apps[ep.Application]; ok {
+					return fmt.Errorf("Duplicate Application detected for endpoint %s found for service %s id %s", ep.Name, svc.Name, svc.ID)
+				}
+			}
+		}
+
+	}
+
+	// If this is a dry run, don't perform the following changes.
+	if request.DryRun {
+		return nil
+	}
+
+	// Migrate the modified services.
+	for _, svc := range svcsMap["modified"] {
+		if err = f.stopServiceForUpdate(ctx, *svc); err != nil {
+			return err
+		}
+		if err = f.migrateService(ctx, svc); err != nil {
+			return err
 		}
 	}
 
-	return err
+	// Wipe out the cloned service's endpoint AddressAssignment data.
+	for _, svc := range svcsMap["cloned"] {
+		for _, ep := range svc.Endpoints {
+			ep.AddressAssignment = addressassignment.AddressAssignment{}
+		}
+	}
+
+	// Add the cloned services.
+	for _, svc := range svcsMap["cloned"] {
```

Comment:
We should verify this with someone who knows the domain better than I do, but I'm guessing that cloned services need to be added BEFORE modified services are migrated. Specifically, if I create a clone and then reparent an existing service to that clone, then I'm guessing the update for the existing service will fail if the clone does not already exist when we migrate the existing service.